File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* findK(ListNode* temp, int k){
        k-=1;
        while(temp!=NULL && k> 0){
            temp = temp->next;
            k--;
        }
        return temp;
    }

    ListNode* reverse(ListNode* temp){
            ListNode* prev = NULL;
            while(temp){
                ListNode * nextNode = temp -> next;
                temp -> next = prev;
                prev= temp;
                temp = nextNode;
            }
            return  prev;
        }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* temp = head;
        ListNode* previous = NULL;
        while(temp){
            ListNode* kth = findK(temp, k);
            if(kth == NULL){
                if (previous) previous -> next = temp;
                break;
            }
            ListNode* nextG = kth->next;
            kth->next = NULL;
            ListNode* headnext = reverse(temp);
            if(head == temp){
                head = kth;
            }else{
                previous -> next = headnext;
            }
            previous = temp;
            temp = nextG;
        }
        return head;        
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        def len(head, k):
            counter= 0
            curr=head
            while(curr):
                curr= curr.next
                counter+=1
            return counter
        N= len(head, k)
        currhead=head
        prevhead=None
        ans=head
        while(N>=k):            
            curr=currhead
            prev=None

            for _ in range(k):
                nxt= curr.next
                curr.next= prev
                prev= curr
                curr=nxt
            
            if prevhead is not None:
                prevhead.next=prev
            else:
                ans=prev
            prevhead=currhead
            currhead=curr
            N=N-k

        if prevhead: 
            prevhead.next = currhead
        return ans